### DISCUSSION TITLE
How does one move a task to a context?

### DISCUSSION THREAD

**Discussion by kingram6865** [GothenburgBitFactory/taskwarrior#2720] [Community support]
I created tasks, then added them to a project, then I found out about contexts and I want to move the tasks in the project to a context I created.

**Comment by smemsh** [comment]
Context is just a shortcut for specifying things like project to all read/write commands used.  The issues themselves don't know anything about context, it's just a way for taskwarrior commands to automatically have arguments added to them, basically.  Are you maybe looking to change the project of a task you already created?

**Reply by kingram6865** [reply]
Thanks for the reply. I did figure out how it works.
I was trying out some things so I deleted and rebuilt my task repository.
Now I understand that the context of a task is defined by the tag used to define the context.
It all makes sense now. Not too complex.

**Reply by smemsh** [reply]
Just to clarify, tags do not "define" context.  A context can include one or more tags, but can also include projects, due date or other things you would normally include on the command line.  Tags are attached to tasks as part of their metadata, whereas context is a "global set of modifiers" (such as tags) that get attached to all invocations of taskwarrior, while under that context.  Hopefully that makes sense.

**Reply by kingram6865** [reply]
Ok. Thanks for the clarification. It would be clear if the documentation offered more info.

From your clarification I am of the understanding that a context is more complex than a tag. It's a set of characteristics that includes many parts, beyond just a tag.

I have to be brutally honest and express my feelings about the documentation. It is really poor I must say.
I'm not sure why, but that doesn't matter. I'm figuring out how to use the tool.

Based on the example I thought my context was defined by +project: A:B:C
When I tagged my task with the identifier I had created for my context, the task landed in the context.

The app is great. The poor documentation made for a frustrating week that I had to spend more time figuring out how to build my project tracking rather than work on tasks.

**Reply by smemsh** [reply]
> poor documentation [about contexts]

Hm, I thought the documentation was good.  Context is just a persistent set of parameters that get added to taskwarrior invocations.  The taskwarrior commands that have their state affected by context are enumerated in the man page (whole sections in both task(1) and taskrc(5) and there's a lot of examples).  There's also https://taskwarrior.org/docs/context.html which seems like a good overview, but hasn't been updated yet since context was given read and write subcontexts in v2.6

There's no need to use context at all, your tasks can't lose or gain anything by using them.  it's only a shortcut to avoid typing the same parameters repeatedly on the command line (such as switching to a work context that adds "+work" tag if you want all your "task add" commands to have the "work" tag while you are at work).

> Based on the example

Based on which example?

> I thought my context was defined by +project: A:B:C

If you add a task and use `+` that's only for specifying tags.  If you used `+project:A:B:C` that would normally filter or add a tag named `project:A:B:C`, not a project named `A:B:C`, but actually since you tried to add a tag with a colon, it would end up in the description instead since colons aren't supported in tags (because they're used to specify attributes like project).  Tags and projects are two different things, and context isn't a thing attached to a task at all, it's only some parameters that get added to an invocation of taskwarrior.

Furthermore, if you're trying to make a hierarchical project, you want to use `.` instead of `:`, although nothing is stopping you from using a project name that includes colons, but this is treated as only a single-level project without any hierarchy.

```
task add project:a.b.c +mytag my description foo bar
```
^^^ this adds a task with project `a`, subproject `b`, subsubproject `c`, tag `mytag`, and description `my description foo bar`.
```
task add project:a:b:c +mytag my description foo bar
```
^^^ this adds a task with project `a:b:c`, tag `mytag`, and description `my description foo bar`.
```
task add +project:a:b:c my description foo bar
```
^^^ this adds a task with description `+project:a:b:c my description foo bar` and no assigned project or tags.

> When I tagged my task with the identifier I had created for my context, the task landed in the context

I think the main confusion could be that context is something stored with tasks.  It's not.  Tasks don't possess context at all; they possess tags, project, etc.
